parquet-variant-compute/src/variant_array.rs

```diff
@@ -647,70 +883,57 @@ fn typed_value_to_variant(typed_value: &ArrayRef, index: usize) -> Variant<'_, '
     }
 }
 
-impl Array for VariantArray {
-    fn as_any(&self) -> &dyn Any {
-        self
-    }
-
-    fn to_data(&self) -> ArrayData {
-        self.inner.to_data()
-    }
-
-    fn into_data(self) -> ArrayData {
-        self.inner.into_data()
-    }
-
-    fn data_type(&self) -> &DataType {
-        self.inner.data_type()
-    }
-
-    fn slice(&self, offset: usize, length: usize) -> ArrayRef {
-        let inner = self.inner.slice(offset, length);
-        let metadata = self.metadata.slice(offset, length);
-        let shredding_state = self.shredding_state.slice(offset, length);
-        Arc::new(Self {
-            inner,
-            metadata,
-            shredding_state,
-        })
-    }
-
-    fn len(&self) -> usize {
-        self.inner.len()
-    }
-
-    fn is_empty(&self) -> bool {
-        self.inner.is_empty()
-    }
-
-    fn offset(&self) -> usize {
-        self.inner.offset()
-    }
-
-    fn nulls(&self) -> Option<&NullBuffer> {
-        self.inner.nulls()
-    }
-
-    fn get_buffer_memory_size(&self) -> usize {
-        self.inner.get_buffer_memory_size()
-    }
+/// Workaround for lack of direct support for BinaryArray
+/// <https://github.com/apache/arrow-rs/issues/8387>
+///
+/// The values are read as
+/// * `StructArray<metadata: Binary, value: Binary>`
+///
+/// but VariantArray needs them as
+/// * `StructArray<metadata: BinaryView, value: BinaryView>`
+///
+/// So cast them to get the right type.
+fn cast_to_binary_view_arrays(array: &dyn Array) -> Result<ArrayRef, ArrowError> {
+    let new_type = map_type(array.data_type());
+    cast(array, &new_type)
+}
 
-    fn get_array_memory_size(&self) -> usize {
-        self.inner.get_array_memory_size()
+/// replaces all instances of Binary with BinaryView in a DataType
+fn map_type(data_type: &DataType) -> DataType {
+    match data_type {
+        DataType::Binary => DataType::BinaryView,
+        DataType::List(field) => {
+            let new_field = field
+                .as_ref()
+                .clone()
+                .with_data_type(map_type(field.data_type()));
+            DataType::List(Arc::new(new_field))
+        }
+        DataType::Struct(fields) => {
+            let new_fields: Fields = fields
+                .iter()
+                .map(|f| {
+                    let new_field = f.as_ref().clone().with_data_type(map_type(f.data_type()));
+                    Arc::new(new_field)
+                })
+                .collect();
+            DataType::Struct(new_fields)
```

Comment:
You _might_ be able to cheat `fmt` by:
```suggestion
            let new_fields = fields.iter().map(|f| {
                Arc::new(f.as_ref().clone().with_data_type(map_type(f.data_type())))
            });
            DataType::Struct(new_fields.collect())
```